crates/provider/src/layers/join_fill.rs

```diff
@@ -0,0 +1,227 @@
+        }
+    }
+}
+
+impl<L, R, N> TxFiller<N> for JoinFill<L, R, N>
+where
+    L: TxFiller<N>,
+    R: TxFiller<N>,
+    N: Network,
+{
+    type Fillable = (Option<L::Fillable>, Option<R::Fillable>);
+
+    fn ready(&self, tx: &N::TransactionRequest) -> bool {
+        self.left.ready(tx) || self.right.ready(tx)
+    }
+
+    fn finished(&self, tx: &N::TransactionRequest) -> bool {
+        self.left.finished(tx) && self.right.finished(tx)
+    }
+
+    async fn request<P, T>(
+        &self,
+        provider: &P,
+        tx: &N::TransactionRequest,
+    ) -> TransportResult<Self::Fillable>
+    where
+        P: Provider<N, T>,
+        T: Transport + Clone,
+    {
+        try_join!(self.left_req(provider, tx), self.right_req(provider, tx))
+    }
+
+    fn fill(&self, to_fill: Self::Fillable, tx: &mut N::TransactionRequest) {
+        if let Some(to_fill) = to_fill.0 {
+            self.left.fill(to_fill, tx);
+        };
+        if let Some(to_fill) = to_fill.1 {
+            self.right.fill(to_fill, tx);
+        };
+    }
+}
+
+impl<L, R, P, N, T> ProviderLayer<P, N, T> for JoinFill<L, R, N>
+where
+    L: TxFiller<N>,
+    R: TxFiller<N>,
+    P: Provider<N, T>,
+    N: Network,
+    T: alloy_transport::Transport + Clone,
+{
+    type Provider = FillProvider<JoinFill<L, R, N>, P, N, T>;
+    fn layer(&self, inner: P) -> Self::Provider {
+        FillProvider::new(inner, self.clone())
+    }
+}
+
+/// A [`Provider`] that applies one or more [`TxFiller`]s.
+///
+/// Fills arbitrary properties in a transaction request by composing multiple
+/// fill layers.
+
+#[derive(Debug, Clone)]
+pub struct FillProvider<F, P, N, T>
+where
+    F: TxFiller<N>,
+    P: Provider<N, T>,
+    N: Network,
+    T: Transport + Clone,
+{
+    inner: P,
+    filler: F,
+    _pd: PhantomData<fn() -> (N, T)>,
+}
+
+impl<F, P, N, T> FillProvider<F, P, N, T>
+where
+    F: TxFiller<N>,
+    P: Provider<N, T>,
+    N: Network,
+    T: Transport + Clone,
+{
+    /// Creates a new `FillProvider` with the given filler and inner provider.
+    pub fn new(inner: P, filler: F) -> Self {
+        Self { inner, filler, _pd: PhantomData }
+    }
+
+    /// Joins a filler to this provider
+    pub fn join_with<Other: TxFiller<N>>(
+        self,
+        other: Other,
+    ) -> FillProvider<JoinFill<F, Other, N>, P, N, T> {
+        self.filler.join_with(other).layer(self.inner)
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<F, P, N, T> Provider<N, T> for FillProvider<F, P, N, T>
+where
+    F: TxFiller<N>,
+    P: Provider<N, T>,
+    N: Network,
+    T: Transport + Clone,
+{
+    fn root(&self) -> &RootProvider<N, T> {
+        self.inner.root()
+    }
+
+    async fn send_transaction(
+        &self,
+        mut tx: N::TransactionRequest,
+    ) -> TransportResult<PendingTransactionBuilder<'_, N, T>> {
+        while self.filler.ready(&tx) && !self.filler.finished(&tx) {
```

Comment:
is this to support hierarchies of fillers?

Reply:
it's to support complex and multi-layer interdependencies. 

i.e. `Signer` fills the `from` then the `NonceManager` fills the nonce while the `GasEstimator` fills the gas, then the `Signer` signs the transaction

Reply:
ah I see.